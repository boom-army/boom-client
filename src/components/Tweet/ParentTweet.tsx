import React, { useContext } from "react";
import { ThemeContext } from "../../contexts/theme";
import { Box } from "@mui/system";
import { Link } from "react-router-dom";
import { UserContext } from "../../contexts/user";

interface Props {
  parentTweet: String | undefined;
}

export const ParentTweet: React.FC<Props> = (props) => {
  const { theme } = useContext(ThemeContext);
  const { user } = useContext(UserContext);

  return (
    <>
      <Box sx={{ background: theme.tertiaryColor, padding: "1rem" }}>
        <Link to={`/${user?.handle}/status/${props.parentTweet}`}>
          Go up thread to parent &#x2934;
        </Link>
      </Box>
    </>
  );
};
