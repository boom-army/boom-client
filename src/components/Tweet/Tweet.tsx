import React from "react";
import "linkify-plugin-hashtag";
import "linkify-plugin-mention";
import Linkify from "linkify-react";
import UserAvatar from "../UserAvatar";
import moment from "moment";
import { CommentIcon } from "../Icons";
import { EmojiTweet, Retweet } from "./index";
import { ImageBox } from "../ImageBox";
import { LAMPORTS_PER_SOL } from "../../constants/math";
import { Link } from "react-router-dom";
import { List as ReactionsList } from "../Reactions/List";
import { NFTTweet } from "../NFT/NFTTweet";
import { TipCreator } from "../TipCreator";
import { styled } from "@mui/material/styles";
import { Tweet } from "../../generated/graphql";
import { VideoContainer } from "../Giphy/VideoContainer";
import { setDate } from "../../utils";
import { useReaction } from "../../hooks/useReaction";

const Wrapper = styled("div")((props) => ({
  display: "flex",
  borderBottom: `1px solid ${props.theme.tertiaryColor}`,
  padding: "1.5rem 1rem 1rem 1rem",

  ".avatar": {
    marginRight: "1em",
    marginBottom: "1em",
  },

  ".tweet-info": {
    width: "100%",
  },

  ".tweet-info-user": {
    display: "flex",
  },

  ".tweet-info-user span.username": {
    fontWeight: 500,
  },

  ".tweet-info-user span.secondary": {
    paddingLeft: "0.5rem",
    color: props.theme.secondaryColor,
  },

  ".tweet-body": {
    marginBottom: "0.75rem",
    wordBreak: "break-word",
  },

  ".tags": {
    display: "flex",
  },

  "a.body": {
    color: props.theme.accentColor,
  },

  "div.tweet-stats": {
    display: "inline-flex",
    flexWrap: "wrap",
    alignItems: "center",

    div: {
      color: props.theme.secondaryColor,
    },

    ".controls": {
      marginRight: "4rem",
    },

    svg: {
      marginRight: "0.5rem",
    },

    span: {
      display: "flex",
      alignItems: "center",
    },

    "span.comment": {
      svg: {
        position: "relative",
        top: "4px",
      },
    },
  },

  "@media screen and (max-width: 470px)": {
    "div.tweet-stats": {
      div: {
        marginRight: "1.5rem",
      },
      ".controls": {
        marginRight: "1.5rem",
      },
    },
  },

  "@media screen and (max-width: 430px)": {
    flexDirection: "column",

    ".username": {
      display: "none",
    },

    ".avatar": {
      display: "none",
    },

    ".tweet-info-user span.secondary": {
      paddingLeft: "0",
      paddingRight: "0.5rem",
    },
  },
}));

interface Props {
  tweet: Tweet;
}

export const ShowTweet: React.FC<Props> = ({ tweet }:Props) => {
  const {
    id,
    text,
    // tags,
    user,
    files,
    gif,
    nft,
    // isTweetMine,
    isRetweet,
    retweetsCount,
    reactions,
    commentsCount,
    tipsCount,
    createdAt,
  } = tweet;

  const { handleReaction } = useReaction({ tweetId: id });
  const handle = user && user.handle;

  const linkifyOptions = {
    className: "body",
    target: { url: "_blank" },
    formatHref: { hashtag: (href: any) => `explore?=${href.substring(1)}` },
  };
  return (
    <Wrapper>
      <Link to={`/${handle}`}>
        <UserAvatar className="avatar" avatar={user?.avatar as string} />
      </Link>

      <div className="tweet-info">
        <div className="tweet-info-user">
          <Link to={`/${handle}`}>
            <span className="username">{user && user.consumerName}</span>
            <span className="secondary">{`@${handle}`}</span>
          </Link>
          &nbsp;&nbsp;
          <Link to={`/${handle}/status/${id}`} className="secondary">
            {moment(setDate(createdAt)).fromNow()}
          </Link>
        </div>

        <Linkify options={linkifyOptions}>
          <p className="tweet-body">{text}</p>
        </Linkify>

        {gif && <VideoContainer gif={gif} />}

        {nft && <NFTTweet nftData={nft} />}

        {!!files.length && <ImageBox files={files} disablelightbox={false} />}

        {reactions.length > 0 && (
          <ReactionsList
            reactions={reactions}
            handleReaction={handleReaction}
            tweetId={id}
          />
        )}

        <div className="tweet-stats">
          <EmojiTweet handleReaction={handleReaction} />

          <div className="controls">
            <span className="comment">
              <Link to={`/${handle}/status/${id}`}>
                <CommentIcon />
                {commentsCount ? commentsCount : null}
              </Link>
            </span>
          </div>

          <div className="controls">
            <Retweet
              id={id}
              isRetweet={isRetweet}
              retweetsCount={retweetsCount}
            />
          </div>

          <>
            <TipCreator
              userPubKey={user?.publicAddress}
              tipAmount={tipsCount && parseInt(tipsCount) / LAMPORTS_PER_SOL}
              tweetId={id}
              userId={user?.id}
            />
          </>

          {/* <div>
            <span>{isTweetMine ? <DeleteTweet id={id} /> : null}</span>
          </div> */}
        </div>
      </div>
    </Wrapper>
  );
};
