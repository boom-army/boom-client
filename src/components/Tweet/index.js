export { DeleteTweet } from './DeleteTweet';
export { EmojiTweet } from './EmojiTweet';
export { MasterTweet } from './MasterTweet';
export { NewTweet } from './NewTweet';
export { ParentTweet } from './ParentTweet';
export { Retweet } from './Retweet';
export { ShowTweet } from './Tweet';
