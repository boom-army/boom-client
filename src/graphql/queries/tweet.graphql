query tweet($id: ID!) {
  tweet(id: $id) {
    ...TweetData
    masterTweets {
      ...TweetData
    }
  }
}

mutation newTweet(
  $text: String!
  $files: [String!]!
  $tags: [String!]!
  $mentions: [String!]!
  $gif: GifInput
  $nft: NFTInput
  $parentTweet: String
  $masterTweet: String
  $channel: String
) {
  newTweet(
    text: $text
    files: $files
    tags: $tags
    mentions: $mentions
    gif: $gif
    nft: $nft
    parentTweet: $parentTweet
    masterTweet: $masterTweet
    channel: $channel
  ) {
    id
    text
    tags
    mentions
    commentsCount
    createdAt
  }
}

mutation toggleRetweet($id: ID!) {
  toggleRetweet(id: $id)
}

mutation deleteTweet($id: ID!) {
  deleteTweet(id: $id) {
    id
  }
}
