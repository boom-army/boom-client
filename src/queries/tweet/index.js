import gql from "graphql-tag";

export const NEW_TWEET = gql`
  mutation newTweet(
    $text: String!
    $files: [String!]!
    $tags: [String!]!
    $mentions: [String!]!
    $gif: GifInput
    $nft: NFTInput
    $parentTweet: String
  ) {
    newTweet(
      text: $text
      files: $files
      tags: $tags
      mentions: $mentions
      gif: $gif
      nft: $nft
      parentTweet: $parentTweet
    ) {
      id
      text
      tags
      mentions
      commentsCount
      createdAt
    }
  }
`;

export const DELETE_TWEET = gql`
  mutation deleteTweet($id: ID!) {
    deleteTweet(id: $id) {
      id
    }
  }
`;

export const TOGGLE_RETWEET = gql`
  mutation toggleRetweet($id: ID!) {
    toggleRetweet(id: $id)
  }
`;

export const TOGGLE_REACTION = gql`
  mutation toggleReaction($id: ID!, $emojiId: String!, $skin: Int) {
    toggleReaction(id: $id, emojiId: $emojiId, skin: $skin)
  }
`;

const TWEET_FRAGMENT = gql`
  fragment subTweet on Tweet {
    id
    text
    tags
    mentions
    retweetsCount
    isRetweet
    createdAt
    parentTweet {
      id
    }
    user {
      id
      publicAddress
      consumerName
      handle
      avatar
    }
    files {
      id
      url
    }
    gif {
      id
      title
      fixedHeightUrl
      originalUrl
    }
    nft {
      id
      publicKey
      name
      symbol
      description
      sellerFeeBasisPoints
      externalUrl
      image
      attributes {
        traitType
        value
      }
      collection {
        name
        family
      }
      properties {
        files {
          uri
          type
        }
        category
        creators {
          address
          share
        }
      }
    }
    reactions {
      id
      emojiId
      skin
      isMine
      count
    }
  }
`;

export const TWEET = gql`
  query tweet($id: ID!) {
    tweet(id: $id) {
      ...subTweet
      childTweets {
        ...subTweet
        childTweets {
          ...subTweet
        }
      }
    }
  }
  ${TWEET_FRAGMENT}
`;
