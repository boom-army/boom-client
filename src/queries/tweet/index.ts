import gql from "graphql-tag";

export const DELETE_TWEET = gql`
  mutation deleteTweet($id: ID!) {
    deleteTweet(id: $id) {
      id
    }
  }
`;

export const TOGGLE_RETWEET = gql`
  mutation toggleRetweet($id: ID!) {
    toggleRetweet(id: $id)
  }
`;

export const TOGGLE_REACTION = gql`
  mutation toggleReaction($id: ID!, $emojiId: String!, $skin: Int) {
    toggleReaction(id: $id, emojiId: $emojiId, skin: $skin)
  }
`;

const TWEET_FRAGMENT = gql`
  fragment subTweet on Tweet {
    id
    text
    tags
    mentions
    retweetsCount
    commentsCount
    isRetweet
    tipsCount
    createdAt
    parentTweet {
      id
    }
    user {
      id
      publicAddress
      consumerName
      handle
      avatar
    }
    files {
      id
      url
    }
    gif {
      id
      title
      fixedHeightUrl
      originalUrl
    }
    nft {
      id
      publicKey
      name
      symbol
      description
      sellerFeeBasisPoints
      externalUrl
      image
      attributes {
        traitType
        value
      }
      collection {
        name
        family
      }
      properties {
        files {
          uri
          type
        }
        category
        creators {
          address
          share
        }
      }
    }
    reactions {
      id
      emojiId
      skin
      isMine
      count
    }
  }
`;

export const TWEET = gql`
  query tweet($id: ID!) {
    tweet(id: $id) {
      ...subTweet
      childTweets {
        ...subTweet
      }
    }
  }
  ${TWEET_FRAGMENT}
`;
