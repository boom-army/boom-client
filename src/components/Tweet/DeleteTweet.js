import React from "react";
import { DELETE_TWEET } from "../../queries/tweet";
import { FEED } from "../../queries/others";
import { TrashIcon } from "../Icons";
import { useMutation } from "@apollo/client";
import { useSnackbar } from "notistack";

export const DeleteTweet = ({ id }) => {
  const [deleteTweetMutation, { loading }] = useMutation(DELETE_TWEET, {
    variables: { id },
    update: (cache, { data: { deleteTweet } }) => {
      const { feed } = cache.readQuery({ query: FEED });
      cache.writeQuery({
        query: FEED,
        data: {
          feed: feed.filter((tweet) => tweet.id !== deleteTweet.id),
        },
      });
    },
  });
  const { enqueueSnackbar } = useSnackbar();

  const handleDeleteTweet = async () => {
    await deleteTweetMutation();
    enqueueSnackbar("Your tweet has been deleted", { variant: "success" });
  };

  return <TrashIcon loading={loading} onClick={handleDeleteTweet} />;
};
