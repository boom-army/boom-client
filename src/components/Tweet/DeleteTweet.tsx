import React from "react";
import { DELETE_TWEET } from "../../queries/tweet";
import { TrashIcon } from "../Icons";
import { useMutation } from "@apollo/client";
import { useSnackbar } from "notistack";
import { FeedDocument, Tweet } from "../../generated/graphql";

export const DeleteTweet = ({ id }: any) => {
  const [deleteTweetMutation, { loading }] = useMutation(DELETE_TWEET, {
    variables: { id },
    update: (cache, { data: { deleteTweet } }) => {
      const { feed }: any = cache.readQuery({ query: FeedDocument });
      cache.writeQuery({
        query: FeedDocument,
        data: {
          feed: feed.filter((tweet: Tweet) => tweet.id !== deleteTweet.id),
        },
      });
    },
  });
  const { enqueueSnackbar } = useSnackbar();

  const handleDeleteTweet = async () => {
    await deleteTweetMutation();
    enqueueSnackbar("Your tweet has been deleted", { variant: "success" });
  };

  return <TrashIcon loading={loading} onClick={handleDeleteTweet} />;
};
