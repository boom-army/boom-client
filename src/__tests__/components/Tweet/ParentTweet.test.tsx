import React from "react";
import { ParentTweet } from "../../../components/Tweet/ParentTweet";
import { TOGGLE_REACTION, TWEET } from "../../../queries/tweet/index";
import { act, fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { createSerializer } from '@emotion/jest'

expect.addSnapshotSerializer(createSerializer(
  {
    classNameReplacer(className, index) {
      return `css-${index}`
    },
       // includeStyles: false 
  }
));


export const __mocks__: any = [
  {
    request: {
      query: TOGGLE_REACTION,
      variables: {},
      refetchQueries: [
        { query: TWEET, variables: { id: "cl084xdy815761s0n5omoxoem0" } },
      ],
    },
    result: {
      data: {
        mentions: [],
      },
      errors: "An error occurred",
    },
  },
];

test("parent tweet snapshot testing", async () => {
  let rendered;
  await act(async () => {
    rendered = render(
      <ParentTweet parentTweet={"cl084xdy815761s0n5omoxoem0"} />,

      { wrapper: MemoryRouter }
    );
  });
  expect(rendered).toMatchSnapshot();
});

test("Assertion testing of <ParentTweet/> component", () => {
  render(<ParentTweet parentTweet={"cl084xdy815761s0n5omoxoem0"} />, {
    wrapper: MemoryRouter,
  });
  expect(
    screen.getByRole("link", { name: "Go up thread to parent â¤´" })
  ).toHaveAttribute("href", "/undefined/status/cl084xdy815761s0n5omoxoem0");
});
