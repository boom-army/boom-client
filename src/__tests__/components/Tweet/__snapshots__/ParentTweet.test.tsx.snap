// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parent tweet snapshot testing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .css-0 {
  background: #38444D;
  padding: 1rem;
}

<body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <a
          href="/undefined/status/cl084xdy815761s0n5omoxoem0"
        >
          Go up thread to parent ⤴
        </a>
      </div>
    </div>
  </body>,
  "container": .css-0 {
  background: #38444D;
  padding: 1rem;
}

<div>
    <div
      class="MuiBox-root css-0"
    >
      <a
        href="/undefined/status/cl084xdy815761s0n5omoxoem0"
      >
        Go up thread to parent ⤴
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
